### ISSUE TITLE
Will the JVM's hotspot compiler sometimes compile to branchless (`CMOVcc`) CPU instructions?

### ISSUE THREAD

**Issue by mikemccand** [Tony-X/search-benchmark-game#43]
One of the cool optimizations Tantivy (and many other places, apparently) use is [branchless algorithms like binary search](https://quickwit.io/blog/search-a-sorted-block).  Another blog post about [branchless binary search](https://probablydance.com/2023/04/27/beautiful-branchless-binary-search/).

It compiles down to CPU instructions called "conditional move", which will move data from one place to another depending on a condition, versus a more general `if` statement that changes which instructions the CPU executes.  `CMOVcc` does not hide memory latency since there is still a data path dependency on both the condition and the source to move, but it does avoid costly branch mis-predictions, which for algos like binary search, or data structures like priority queues, can be frequent.

This got me wondering: would such implementations in Java also compile down to `CMOVcc` instructions?

I tried googling and didn't find any clear answers.  There are some indications, like [this JDK bug](https://bugs.openjdk.org/browse/JDK-8039104), but that is moving in the opposite direction (NOT compiling down to `CMOVcc`).

Is there somehow a chance for branchless algorithms and data structures to work well in Java too?

**Comment by Tony-X** [comment]
It will be super interesting to examine how JVM optimize the hot paths like sorting and searching within a block.

is there a way to access the compiled native code associated with a method?

**Comment by mikemccand** [comment]
> is there a way to access the compiled native code associated with a method?

Yes there is!  https://blogs.oracle.com/javamagazine/post/java-hotspot-hsdis-disassembler

The Lucene devs used this heavily when trying to get HotSpot to auto-vectorize down to SIMD by writing the Java sources "just right".
